File: Python/LinkedList/Palindrome LinkedList.py
//https://www.geeksforgeeks.org/problems/check-if-linked-list-is-pallindrome/1
def reverse(self, head):
        prev = None
        current = head
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        return prev

    def checkpalindrome(self, head, head1):
        while head and head1:
            if head.data != head1.data:
                return False
            head = head.next
            head1 = head1.next
        return True

    def isPalindrome(self, head):
        if not head:
            return True
        
        # Find the middle of the list
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # Reverse the second half of the list
        second_half = self.reverse(slow)
        
        # Check if the first half and reversed second half are the same
        return self.checkpalindrome(head, second_half)

File: Java/Linked List/Palindrome Linked List.java
//https://www.geeksforgeeks.org/problems/check-if-linked-list-is-pallindrome/1
class Solution {
    // Function to check whether the list is palindrome.
    
    public static Node reverse(Node head) {
        Node temp = null;
        Node temp1 = head;
        Node temp2;

        while (temp1 != null) {
            temp2 = temp1.next;
            temp1.next = temp;
            temp = temp1;
            temp1 = temp2;
        }
        return temp;
    }

    public static boolean checkPalindrome(Node head, Node head1) {
        while (head != null && head1 != null) {
            if (head.data != head1.data) {
                return false;
            }
            head = head.next;
            head1 = head1.next;
        }
        return true;
    }
    static boolean isPalindrome(Node head) {
        // Your code here
        if (head == null || head.next == null) return true;

        Node temp = head;
        Node temp1 = head.next;

        while (temp1 != null && temp1.next != null) {
            temp = temp.next;
            temp1 = temp1.next.next;
        }

        Node temp2 = temp.next;
        temp.next = null;

        Node newHead = reverse(temp2);

        return checkPalindrome(head, newHead);
    }
}
